### ISSUE TITLE
Non-modules allowed in module positions in `use`.

### ISSUE THREAD

**Issue by SergioBenitez** [rust-lang/rust#38293]
Consider the following piece of code, which compiles:

```rust
mod bar {
    pub fn foo() { }
}

use bar::foo::{self};

fn main() {
    foo();
}
```

The `use` declaration is allowed to use `foo` as a module when `foo` is actually a function. This should not be allowed.

Further, when `foo` _does_ exist as a module, regardless of visibility, `self` incorrectly imports items with the same name in that module:

```rust
mod bar {
    mod foo { }
    pub fn foo() { }
}

use bar::foo::{self};

fn main() {
    foo();
}
```

**Comment by jseyfried** [comment]
cc @petrochenkov

**Comment by jseyfried** [comment]
Fixed in #38313.
